core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala

```diff
@@ -0,0 +1,156 @@
+  def getExecutorBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String]
+
+  def getNodeBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String]
+
+  // Default implementation to remove failure executors from HashMap based on given time period.
+  def expireExecutorsInBlackList(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]): Unit = {
+    val now = new SystemClock().getTimeMillis()
+    failedExecutorMap.retain((executorid, failureStatus) => {
+      (now - failureStatus.updatedTime) < expireTimeInMillisecond
+    })
+  }
+}
+
+/**
+ * The interface for taskId unrelated strategy. In this case, provided taskId will be ignored. The
+ * benefit for taskId unrelated strategy is different taskSets can learn experience from other
+ * taskSet to avoid allocating tasks to problematic executors.
+ */
+trait TaskUnrelatedBlacklistStrategy extends BlacklistStrategy with Logging{
+  def getExecutorBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String] = {
+    if (taskId.isDefined){
+      logWarning("TaskId unrelated strategy is applied, given taskId will be ignored")
+    }
+    getExecutorBlacklist(failedExecutorMap)
+  }
+
+  def getNodeBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String] = {
+    if (taskId.isDefined){
+      logWarning("TaskId unrelated strategy is applied, given taskId will be ignored")
+    }
+    getNodeBlacklist(failedExecutorMap)
+  }
+
+  // define new interface which remove taskId from parameter.
+  def getExecutorBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]): Set[String]
+  def getNodeBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]): Set[String]
+}
+
+/**
+ * This strategy is simply based on given threshold and is taskId unrelated. An executor will be
+ * in blacklist, if it failed more than "maxFailureTaskNumber" times. A node will be in blacklist,
+ * if there are more than "maxBlackExecutorNumber" executors on it in executor blacklist.
+ */
+class SimpleStrategy(
+    maxFailureTaskNumber: Int,
+    maxBlackExecutorNumber: Int,
+    val expireTimeInMillisecond: Long
+  )extends TaskUnrelatedBlacklistStrategy {
+
+  private def getSelectedExecutorMap(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]) = {
+    failedExecutorMap.filter{
+      case (id, failureStatus) => failureStatus.failureTimes > maxFailureTaskNumber
+    }
+  }
+
+  def getExecutorBlacklist(
+      failureExecutors: mutable.HashMap[String, FailureStatus]): Set[String] = {
+    getSelectedExecutorMap(failureExecutors).keys.toSet
+  }
+
+  def getNodeBlacklist(
+      failureExecutors: mutable.HashMap[String, FailureStatus]): Set[String] = {
```

Comment:
Current thinking is maintain extra "blacklistExecutors" and "blacklistNodes" map as cache and add a AtomicBoolean mark "isChanged" to decide if we need to refresh cache. 

"isChanged" will be set to true when failureExecutors updated. And when we call getExecutorBlacklist, it will check "isChanged" if true, refresh cache, set "isChanged" to false and return, If false, direct return the value in cache.

I am now implementing it, any comments, let me know. Thanks